src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/net/SubscriptionManager.java

```diff
@@ -110,30 +70,36 @@ public SubscriptionStatus addSubscriber(
 
     Set<String> currentSubscribers = subscriberMap.getOrDefault(graphNode, new HashSet<>());
     currentSubscribers.add(subscriberHostAddress);
+    // happens-before: update to a java.util.concurrent collection. Updated value will be visible
+    // to subsequent reads.
     subscriberMap.put(graphNode, currentSubscribers);
 
     LOG.debug("locus matched. Added subscriber {} for {}", subscriberHostAddress, graphNode);
     return SubscriptionStatus.SUCCESS;
   }
 
   public boolean isNodeSubscribed(final String graphNode) {
+    // happens-before: reading from a java.util.concurrent collection which guarantees read
+    // reflects most recent completed update.
     return subscriberMap.containsKey(graphNode);
   }
 
-  public Set<String> getSubscribersFor(final String graphNode) {
-    return subscriberMap.getOrDefault(graphNode, new HashSet<>());
+  public ImmutableSet<String> getSubscribersFor(final String graphNode) {
+    // happens-before: ImmutableSet - final field semantics. Reading from java.util.concurrent
+    // collection.
+    return ImmutableSet.copyOf(subscriberMap.getOrDefault(graphNode, new HashSet<>()));
   }
 
-  private void addPublisher(final String graphNode, final String publisherHostAddress) {
-    LOG.debug("Added publisher: {} for graphNode: {}", publisherHostAddress, graphNode);
+  public synchronized void addPublisher(final String graphNode, final String publisherHostAddress) {
+    LOG.info("Added publisher: {} for graphNode: {}", publisherHostAddress, graphNode);
 
     final Set<String> currentPublishers = publisherMap.getOrDefault(graphNode, new HashSet<>());
     currentPublishers.add(publisherHostAddress);
     publisherMap.put(graphNode, currentPublishers);
   }
 
   public void dumpStats() {
-    LOG.debug("Subbscribers: {}", subscriberMap);
+    LOG.debug("Subscribers: {}", subscriberMap);
```

Comment:
For dumpStats, should we also say the amount of payload each queue has ?

Reply:
Removed dumpstats. Was debug code.